In [1]:
#Import dataset from Kaggle

import pandas as pd

df = pd.read_csv("ransomware.csv")

In [2]:

#Analyzing dataset
print(df.head())
print(df.info())

                                            FileName  \
0           0124e21d-018c-4ce0-92a3-b9e205a76bc0.dll   
1       05c8318f98a5d301d80000009c316005.vertdll.dll   
2           06054fba-5619-4a86-a861-ffb0464bef5d.dll   
3        075822ac99a5d301660400009c316005.adhapi.dll   
4  090607dd9ba5d301ca0900009c316005.SensorsNative...   

                            md5Hash  Machine  DebugSize  DebugRVA  \
0  79755c51e413ed3c6be4635fd729a6e1      332          0         0   
1  95e19f3657d34a432eada93221b0ea16    34404         84    121728   
2  85c32641d77a54e19ba8ea4ab305c791      332          0         0   
3  62e3b959d982ef534b66f819fe15f085    34404         84     19904   
4  ae38c5f7d313ad0ff3bfb8826476767f    34404         84     97728   

   MajorImageVersion  MajorOSVersion  ExportRVA  ExportSize  IatVRA  \
0                  0               4          0           0    8192   
1                 10              10     126576        4930       0   
2                  0               

- By analyzing the structure of the dataset I decided to drop FileName and md5Hash from feature set. 
- They serve as identifiers and the model would just memorize, not generalize

In [3]:
df.describe()

Machine     DebugSize      DebugRVA  MajorImageVersion  \
count  62485.000000  6.248500e+04  6.248500e+04       62485.000000   
mean    6756.472657  2.587048e+04  1.541611e+05          58.785997   
std    13345.499919  6.461396e+06  1.903142e+06        1114.068244   
min        0.000000  0.000000e+00  0.000000e+00           0.000000   
25%      332.000000  0.000000e+00  0.000000e+00           0.000000   
50%      332.000000  0.000000e+00  0.000000e+00           0.000000   
75%      332.000000  2.800000e+01  1.283200e+04           6.000000   
max    43620.000000  1.615155e+09  2.852127e+08       63325.000000   

       MajorOSVersion     ExportRVA    ExportSize        IatVRA  \
count    62485.000000  6.248500e+04  6.248500e+04  6.248500e+04   
mean         5.417524  8.953186e+05  4.094623e+05  1.466311e+05   
std          2.543697  3.779527e+07  2.851820e+07  1.124630e+06   
min          0.000000  0.000000e+00  0.000000e+00  0.000000e+00   
25%          4.000000  0.000000e+00  0.000000e+00  4.096000e+03   
50%          5.000000  0.000000e+00  0.000000e+00  8.520000e+03   
75%          6.000000  2.875200e+04  1.040000e+02  6.553600e+04   
max        260.000000  2.147484e+09  2.415919e+09  6.615450e+07   

       MajorLinkerVersion  MinorLinkerVersion  NumberOfSections  \
count        62485.000000        62485.000000      62485.000000   
mean             9.782604            6.613171          4.751172   
std              9.047157           16.976465          2.138584   
min              0.000000            0.000000          1.000000   
25%              6.000000            0.000000          3.000000   
50%              9.000000            0.000000          5.000000   
75%             11.000000           10.000000          6.000000   
max            255.000000          255.000000         33.000000   

       SizeOfStackReserve  DllCharacteristics  ResourceSize  BitcoinAddresses  \
count        6.248500e+04        62485.000000  6.248500e+04      62485.000000   
mean         8.759830e+05        12966.031400  1.844664e+05          0.018004   
std          6.288189e+05        15762.725511  1.732625e+07          0.132968   
min          0.000000e+00            0.000000  0.000000e+00          0.000000   
25%          2.621440e+05            0.000000  1.080000e+03          0.000000   
50%          1.048576e+06          320.000000  2.496000e+03          0.000000   
75%          1.048576e+06        32768.000000  2.350400e+04          0.000000   
max          1.677722e+07        58632.000000  4.294942e+09          1.000000   

             Benign  
count  62485.000000  
mean       0.433992  
std        0.495628  
min        0.000000  
25%        0.000000  
50%        0.000000  
75%        1.000000  
max        1.000000

## Data Interpretation:

### What each row means

- count → how many non-missing values in the column (all ≈ 62,485, so no missing data).
- mean → the average value across samples.
- std → standard deviation (spread of the values).
- min → smallest observed value.
- 25% / 50% / 75% → quartiles (median = 50%). These tell you the distribution shape.
- max → largest observed value.

### Interesting Rows

* NumberOfSections
    - Normal Windows executables usually have 4–7 sections.
    - Malware sometimes adds extra sections fo payload, packing or obfuscation
    - **good discriminator feature**

* DllCharacteristics
    - Bitmask, where each bit(flag) shows support of requirement of features.
    - Benign software: usually has common protections enabled (ASLR, DEP).
    - Malware may use unusual combinations, or omit common protections.
    - Giant numbers can indicate nonstandard or obfuscated files. 

* ResourceSize
    - Legit exes often have icons, dialogs, etc.
    - Malware may have abnormally large or missing resource sections.
    - Outliers here are interesting.

* Machine
    - Most benign files = 332 (Intel 386).
    - Outliers (weird architectures) could signal malicious or malformed files.
    - Worth keeping.
